File: 37-sudoku-solver.py
from typing import List


class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        '''
        We can use backtracking to solve this
        '''
        row_list = [set() for _ in range(9)]
        col_list = [set() for _ in range(9)]
        square_list = [set() for _ in range(9)]

        def get_square_idx(row, col):
            temp = row // 3
            temp2 = col // 3
            return temp * 3 + temp2

        def check_valid(row, col, val):
            square = get_square_idx(row, col)
            if val in row_list[row] or val in col_list[col] or val in square_list[square]:
                return False

            return True

        def populate_lists():
            for row in range(9):
                for col in range(9):
                    val = board[row][col]
                    if val == '.':
                        continue
                    row_list[row].add(val)
                    col_list[col].add(val)
                    square_list[get_square_idx(row, col)].add(val)

        populate_lists()

        def solve():
            # Find all empty cells once
            empty_cells = []
            for row in range(9):
                for col in range(9):
                    if board[row][col] == '.':
                        empty_cells.append((row, col))

            def backtrack(index):
                if index == len(empty_cells):
                    return True

                row, col = empty_cells[index]
                for i in range(1, 10):
                    val_str = str(i)
                    if check_valid(row, col, val_str):
                        # Place value
                        board[row][col] = val_str
                        row_list[row].add(val_str)
                        col_list[col].add(val_str)
                        square_list[get_square_idx(row, col)].add(val_str)

                        if backtrack(index + 1):
                            return True

                        # Backtrack
                        board[row][col] = '.'
                        row_list[row].remove(val_str)
                        col_list[col].remove(val_str)
                        square_list[get_square_idx(row, col)].remove(val_str)

                return False

            return backtrack(0)

        solve()


sol = Solution()
board = [
    ["5", "3", ".", ".", "7", ".", ".", ".", "."],
    ["6", ".", ".", "1", "9", "5", ".", ".", "."],
    [".", "9", "8", ".", ".", ".", ".", "6", "."],
    ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
    ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
    ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
    [".", "6", ".", ".", ".", ".", "2", "8", "."],
    [".", ".", ".", "4", "1", "9", ".", ".", "5"],
    [".", ".", ".", ".", "8", ".", ".", "7", "9"]]
sol.solveSudoku(board)
print(f'output: {board}')

File: 37-sudoku-solver.java
import java.util.*;

class Cell {
    int row, col;

    public Cell(int row, int col) {
        this.row = row;
        this.col = col;
    }
}

class SudokuSolverSolution {

    public int getSquareIndex(int row, int col) {
        return 3 * (row / 3) + (col / 3);
    }

    public boolean checkValid(int row, int col, char value, List<Set<Character>> rowSets, List<Set<Character>> colSets,
            List<Set<Character>> squareSets) {
        int squareIdx = getSquareIndex(row, col);
        Set<Character> rowSet = rowSets.get(row);
        Set<Character> colSet = colSets.get(col);
        Set<Character> squareSet = squareSets.get(squareIdx);
        if (rowSet.contains(value) || colSet.contains(value) || squareSet.contains(value)) {
            return false;
        }
        return true;
    }

    public void populate(char[][] board, List<Set<Character>> rowSets, List<Set<Character>> colSets,
            List<Set<Character>> squareSets) {
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                int squareIdx = getSquareIndex(r, c);
                Set<Character> rowSet = rowSets.get(r);
                Set<Character> colSet = colSets.get(c);
                Set<Character> squareSet = squareSets.get(squareIdx);

                char value = board[r][c];
                if (value == '.') {
                    continue;
                }

                rowSet.add(value);
                colSet.add(value);
                squareSet.add(value);
            }
        }
    }

    public void solve(char[][] board, List<Set<Character>> rowSets, List<Set<Character>> colSets,
            List<Set<Character>> squareSets) {
        List<Cell> emptyCells = new ArrayList<>();
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (board[r][c] == '.') {
                    emptyCells.add(new Cell(r, c));
                }
            }
        }

        backtrack(board, rowSets, colSets, squareSets, emptyCells, 0);
    }

    public boolean backtrack(char[][] board, List<Set<Character>> rowSets, List<Set<Character>> colSets,
            List<Set<Character>> squareSets, List<Cell> emptyCells, int index) {
        if (index == emptyCells.size()) {
            return true;
        }

        Cell cell = emptyCells.get(index);
        for (int i = 1; i < 10; i++) {
            int row = cell.row;
            int col = cell.col;
            char value = (char) (i + '0');
            if (checkValid(row, col, value, rowSets, colSets, squareSets)) {
                board[row][col] = value;
                int squareIdx = getSquareIndex(row, col);
                Set<Character> rowSet = rowSets.get(row);
                Set<Character> colSet = colSets.get(col);
                Set<Character> squareSet = squareSets.get(squareIdx);
                rowSet.add(value);
                colSet.add(value);
                squareSet.add(value);
                if (backtrack(board, rowSets, colSets, squareSets, emptyCells, index + 1)) {
                    return true;
                }

                board[row][col] = '.';
                rowSet.remove(value);
                colSet.remove(value);
                squareSet.remove(value);
            }
        }
        return false;
    }

    public void solveSudoku(char[][] board) {
        List<Set<Character>> rowSets = new ArrayList<>();
        List<Set<Character>> colSets = new ArrayList<>();
        List<Set<Character>> squareSets = new ArrayList<>();
        for (int i = 0; i < 9; i++) {
            rowSets.add(new HashSet<>());
            colSets.add(new HashSet<>());
            squareSets.add(new HashSet<>());
        }

        populate(board, rowSets, colSets, squareSets);

        solve(board, rowSets, colSets, squareSets);
    }

    public static void main(String[] args) {
        SudokuSolverSolution sol = new SudokuSolverSolution();
        char[][] board;

        board = new char[][] { { '5', '3', '.', '.', '7', '.', '.', '.', '.' },
                { '6', '.', '.', '1', '9', '5', '.', '.', '.' }, { '.', '9', '8', '.', '.', '.', '.', '6', '.' },
                { '8', '.', '.', '.', '6', '.', '.', '.', '3' }, { '4', '.', '.', '8', '.', '3', '.', '.', '1' },
                { '7', '.', '.', '.', '2', '.', '.', '.', '6' }, { '.', '6', '.', '.', '.', '.', '2', '8', '.' },
                { '.', '.', '.', '4', '1', '9', '.', '.', '5' }, { '.', '.', '.', '.', '8', '.', '.', '7', '9' } };
        sol.solveSudoku(board);

        System.out.println(String.format(
                "Output: %s, expected: [['5','3','4','6','7','8','9','1','2'],['6','7','2','1','9','5','3','4','8'],['1','9','8','3','4','2','5','6','7'],['8','5','9','7','6','1','4','2','3'],['4','2','6','8','5','3','7','9','1'],['7','1','3','9','2','4','8','5','6'],['9','6','1','5','3','7','2','8','4'],['2','8','7','4','1','9','6','3','5'],['3','4','5','2','8','6','1','7','9']]",
                Arrays.deepToString(board)));
    }
}
